### PR TITLE
Appium2

### PR DESCRIPTION
**Pull request by burtonemily**
### Move most of the calls to the DeviceWrapper class
Now that you have an abstraction better have a single source of calls you can make. The calls below (at least can be moved) to the DeviceWrapper class directly.
```
   * waitForElementToBePresent
   * waitForTextElementToBePresent
   * waitForXPathElement
   * swipeLeft
   * sendNewMessage
   * sendMessageTo
   * sendMessage
   * const scrollDown
   * const replyToMessage
   * longPressMessage
   * clickOnElement
   * grabTextFromAccessibilityId
   * navigateBack and make it behave the right way based on the current platform type
   * move `runOnlyOniOS` and `Android` to that `DeviceWrapper` class too, no need for the extra argument, you can just use `this.isIos()` or `this.isAndroid()` to know what you are running on. Also, see Note1 below
   * all of the `element_selection.ts` file
   * deleteText
   * I'd say probably everything having device: DeviceWrapper as a first argument but not too sure

```

The DeviceWrapper class might become quite big, but I'd say that's fine and it makes sense

### Make sure only one call doing one thing exists in that DeviceWrapper class

i..e: are `sendMessage`, `sendMessageTo` and `sendNewMessage` not the same? if not exactly, can they be merged together with one extra defaulted argument?

### Force the types allowed by each of of your custom made functions, depending on if they take a string or an accessibilityId

accessibility ids are fixed and known, they cannot change without an android/ios application change.
Fixing they type means you won't be able to make mistakes when adding a new test, because the accessibility ids are defined and you get compile error if you are trying to use which does not exists.

What I mean is to create a file name `accessibility_ids.ts` with the content like:
```ts
export type AccessibilityId = "Conversation list item" | "Message sent status: Sent" | "Reply to message" | "Back" | "Navigate up" | "New conversation button"...;

// You could even break those down into the android and ios accessibility ids
```

make all the functions taking an accessibilityId as argument use that type. 
Essentially search for at least `  accessibilityId: string` and rename to `  accessibilityId: AccessibilityId`
Then fix the errors by adding the missing ids in the `AccessibilityId` union

### Note1
rather than having runOnlyOnOios and Android, you could make all your DeviceWrapper function take an extra defaulted argument which would be the device to run that command on. By default, that would be all devices
```ts
class DeviceWrapper {
[...]
public async waitForTextElementToBePresent (
  {
   accessibilityId, 
   text,
   maxWait,
   devicesToRunOn
} :
{ 
  accessibilityId: string,
  text: string,
  maxWait?: number
  devicesToRunOn: Set<SupportedPlatform> = new Set(["android", "ios"])
  }
)
```

### Later
* Find a way to merge openApp, openAppTwo, Three and for devices together (generic)
* Get rid of all the remaining `: any` in the app (not so easy to do it right)
* Find a way to not have to call openAppXdevices before each test and close it. Use mocha fixtures\
* Make a higher level merged iosIt + androidIt as everything seems to be running for each (not sure if possible that easily, but essentially have a allIt() which define an android test and an ios one (by calling inside androidIt and iosIt)

### CODE REVIEW

File: appium_next.d copy.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+"use strict";
```
Comment (Bilb) [inline appium_next.d copy.js:1]: this file can be deleted completely

File: run/test/specs/disappearing_messages.spec.ts [hunk 2]
```diff
@@ -46,7 +46,8 @@ async function disappearingMessages(platform: SupportedPlatformsType) {
   const message = "Howdy testing disappearing messages";
   await inputText(device1, "Message input box", message);
   // Wait 5 seconds
-  await device1.setImplicitWaitTimeout(5000);
+  console.warn("FIXME audric");
```
Comment (Bilb) [inline run/test/specs/disappearing_messages.spec.ts:49]: to remove

File: run/test/specs/create_user_manual.spec.ts [hunk 3]
```diff
@@ -8,25 +8,25 @@ import {
...
   const [userA, userB] = await Promise.all([
     newUser(device1, "User A", platform),
     newUser(device2, "User B", platform),
   ]);
 
-  await newContact(device1, userA, device2, userB);
-
-  await device1.setImplicitWaitTimeout(50000);
+  await newContact(platform, device1, userA, device2, userB);
+  console.warn("FIXME audric");
```
Comment (Bilb) [inline run/test/specs/create_user_manual.spec.ts:19]: to remove

File: run/test/specs/create_user_manual.spec.ts [hunk 4]
```diff
@@ -8,25 +8,25 @@ import {
...
+  // await device1.setImplicitWaitTimeout(50000);
   console.log("Waiting for manual tests");
-
-  await device2.setImplicitWaitTimeout(50000);
+  console.warn("FIXME audric");
+  // await device2.setImplicitWaitTimeout(50000);
```
Comment (Bilb) [inline run/test/specs/create_user_manual.spec.ts:23]: to remove

**Review by Bilb — CHANGES_REQUESTED** [review]

File: run/test/specs/utils/open_app.ts [hunk 5]
```diff
@@ -111,86 +95,72 @@ export const openAppFourDevices = async (
...
-  const newServer: AppiumServer =
-    server || (await openAppiumServerOnly("android"));
-
   await installAppToDeviceName(
     androidCapabilities.androidAppFullPath,
     getAndroidUuid(capabilitiesIndex)
   );
+  const driver = (androidDriver as any).AndroidUiautomator2Driver;
+
+  // console.warn('installAppToDeviceName ', driver);
```
[the remark this answers is not in the document]
Comment (burtonemily) [reply run/test/specs/utils/open_app.ts:108]: Remove lines 106-108?

File: run/test/specs/utils/check_colour.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import PNG from "png-js";
+
+export async function parseDataImage(base64: string) {
+  console.warn("base64", base64);
```
[the remark this answers is not in the document]
Comment (burtonemily) [reply run/test/specs/utils/check_colour.ts:4]: just the console.warn...?

File: run/test/specs/utils/do_function_if_element_exists.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { DeviceWrapper } from "../../../types/DeviceWrapper";
+
+export const doFunctionIfElementExists = async (
+  device: DeviceWrapper,
+  strategy: "xpath" | "accessibility id",
+  selector: string,
+  toRun: () => Promise<any>
+) => {
+  try {
+    await device.findElement(strategy, selector);
```
[the remark this answers is not in the document]
Comment (burtonemily) [reply run/test/specs/utils/do_function_if_element_exists.ts:10]: Deleted this function, it wasn't doing anything

File: run/test/specs/utils/do_function_if_element_exists.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    // Check if element exists
+    if (selector) {
+      // IF yes do the thing
+      console.log(`'Found ${strategy}: ${selector} so doing the thing`);
+      await toRun();
+    }
+  } catch (e) {
+    console.log(`No ${selector} so moving on`);
+  }
```
[the remark this answers is not in the document]
Comment (burtonemily) [reply run/test/specs/utils/do_function_if_element_exists.ts:20]: Deleted function

File: run/types/sessionIt.ts [hunk 9]
```diff
@@ -1,19 +1,39 @@
...
+
+// async function itWithBufferHandling(testNameWithoutPlatform: string, platform: SupportedPlatformsType, testToRun: () => Promise<void>) {
+//   try {
+//     await testToRun();
+//     clearBufferOfTest(testNameWithoutPlatform);
+//   } catch (e) {
+//     printBufferAndClear(testNameWithoutPlatform)
+//     throw e;
+//   }
+// }
```
Comment (Bilb) [inline run/types/sessionIt.ts:15]: to be removed

File: 21:00:00
Comment (Bilb) [inline 21:00:00:1]: same for this file

**Review by Bilb — APPROVED** [review]

**Pull request merged**